models/tests/utils/test_compile.py

```diff
@@ -144,8 +143,67 @@ def test_compile_layer_kernel() -> None:
     result = mhsa.forward(x, shapes, batch_size)
     result_compiled = mhsa_compiled.forward(x, shapes, batch_size)
 
-    # check the function was compiled
-    assert hasattr(mhsa_compiled.projection.forward, "_torchdynamo_orig_callable")
-
     # check the result of the compiled function matches the uncompiled result
     assert torch.allclose(result, result_compiled)
+
+
+def test_compile_save_checkpoint() -> None:
+    """Tests that a compiled module can be pickled and saved as a checkpoint"""
+    # Skip this test if library versions aren't met
+    if not _meets_library_versions_for_compile():
+        LOGGER.warning("triton not installed. skipping 'test_compile.py::test_compile_save_checkpoint'")
+        return
+
+    num_channels = 64
+    cond_shape = 16
+    ln = ConditionalLayerNorm(num_channels, condition_shape=cond_shape)
+    x_in = torch.randn(num_channels)
+    cond = torch.randn(cond_shape)
+
+    cfg = graphtransformer_ens_compile_config()
+    ln_compiled = mark_for_compilation(ln, cfg.compile)
+
+    # we dont care about the result, but we need to compute it to trigger compilation
+    _ = ln_compiled.forward(x_in, cond)
+
+    # try save checkpoint
+    buffer = io.BytesIO()
+    torch.save(ln_compiled, buffer)
+
+
+def test_compile_load_checkpoint() -> None:
+    """Tests that a compiled checkpoint can be loaded by a non-compiled model"""
+    # Skip this test if library versions aren't met
+    if not _meets_library_versions_for_compile():
+        LOGGER.warning("triton not installed. skipping 'test_compile.py::test_compile_load_checkpoint'")
+        return
+
+    cfg = layer_kernel_compile_config()
+    layer_kernels = load_layer_kernels(kernel_config={})
+
+    num_heads = 1
+    embed_dim = 64
+    dropout_p = 0.0
+    batch_size = 1
+    mhsa = MultiHeadSelfAttention(
+        num_heads,
+        embed_dim,
+        layer_kernels,
+        dropout_p=dropout_p,
+        attention_implementation="scaled_dot_product_attention",
+    )
+    mhsa_compiled = mark_for_compilation(mhsa, cfg.compile)
+
+    x = torch.randn(batch_size * 2, embed_dim, requires_grad=True)
+    shapes = [list(x.shape)]
+
+    result_compiled = mhsa_compiled.forward(x, shapes, batch_size)
+
+    torch.save(mhsa_compiled, "compiled.pt")
+
+    checkpoint = torch.load("compiled.pt", weights_only=False)
+    os.remove("compiled.pt")
+    mhsa.load_state_dict(checkpoint.state_dict(), assign=False)
```

Comment:
in `mark_for_compilation` we seem to be doing the operations in place return the same model as input with the modules compiled. Would it be safer then for this test to use a fresh instantiation of the mhsa for this?

so 
```
    mhsa_new = MultiHeadSelfAttention(Add a comment on  lines R188 to R189Add diff commentMarkdown input:  edit mode selected.WritePreviewAdd a suggestionHeadingBoldItalicQuoteCodeLinkUnordered listNumbered listTask listMentionReferenceSaved repliesAdd FilesPaste, drop, or click to add filesCancelCommentStart a reviewReturn to code
        num_heads,
        embed_dim,
        layer_kernels,
        dropout_p=dropout_p,
        attention_implementation="scaled_dot_product_attention",
    )

mhsa_new.load_state_dict(checkpoint.state_dict(), assign=False)

```